### DISCUSSION TITLE
How can I reproduce the models accuracy evaluations?

### DISCUSSION THREAD

**Discussion by vmatter** [explosion/spaCy#11030] [Help: Coding & Implementations] [answered]
I would like to reproduce the accuracy evaluations of the [Portuguese models ](https://spacy.io/models/pt).

I tried to search for the code used in the accuracy evaluation, however, I could not find it...

Is there a way to reproduce them?

Many thanks.

**Comment by polm** [comment, marked as answer]
Evaluation metrics for a pipline are saved in `meta.json` under the `performance` key. Evaluation (on the dev set) is performed automatically during training, but you can also do it manually (on any data, such as test data) with [`spacy evaluate`](https://spacy.io/api/cli#evaluate). The metrics on the model pages are for the dev set.

**Reply by vmatter** [reply]
Where can I find this `performance` key?

Just to make sure that we are on the same page, I want to know how to reproduce the same results as the one found in the [Portuguese models](https://spacy.io/models/pt):

![accuracy_evaluation](https://user-images.githubusercontent.com/43481916/175817136-6d9dc2f3-4ece-431a-b8b5-13e1970db0e4.png)

My custom model uses the [`spacy evaluate`](https://spacy.io/api/cli#evaluate). 

However, I would really like to know how can I get the general accuracy of the [UD Portuguese Bosque v2.8](https://github.com/UniversalDependencies/UD_Portuguese-Bosque) and [WikiNER](https://figshare.com/articles/Learning_multilingual_named_entity_recognition_from_Wikipedia/5462500) corpus that achieved the F-Score of 0.88.

Did spaCy use these corpus' training and test datasets in the [`spacy evaluate`](https://spacy.io/api/cli#evaluate) function? If so do you know how to get them?

Thank you.

**Reply by polm** [reply]
> Where can I find this performance key?

In the root directory of your pipeline, there is a `meta.json` file. That is a JSON dict, and it has a `performance` key. It looks like this:

```
...
  "performance":{
    "token_acc":0.9993092439,
    "token_p":0.9956819193,
    "token_r":0.9957659295,
    "token_f":0.9957239226,
...
```

Thoe are the numbers on the models page, and are based on an evaluation of the dev set used in training. 

In this case the evaluation is based on two datasets, the UD dataset and WikiNER. I believe in this case the pipelines are trained separately (WikiNER for NER, UD for the rest) and then combined, so you should be able to do the same for evaluation.

If you have the same annotations on your data I think `spacy evaluate` should provide you with the same kinds of scores. 

> Did spaCy use these corpus' training and test datasets in the [spacy evaluate](https://spacy.io/api/cli#evaluate) function? If so do you know how to get them?

For UD corpora, the data is split into train/dev/test, with each in a separate file. So `pt_bosque-ud-dev.conllu` is the dev set that the evaluation numbers are against.

**Reply by vmatter** [reply]
Many thanks, now I got the essence of how it was evaluated.

I will replicate the tests and evaluate the results.

Once again, thank you!
